### PR TITLE
Transform v1 fix, fileChange, replacement

### PR DESCRIPTION
**Pull request by EasyRhin0**
Please check on how I'm transforming arguments that correspond to rule messages, I'm a bit confused about how that works.

### CODE REVIEW

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 1]
```diff
@@ -1061,6 +1061,11 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
         {
           ""ruleId"": ""C2001"",
           ""level"": ""error"",
+          ""message"": {
+            ""arguments"": [
+              ""ptr""
+            ]
+          },
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: This is correct. The spec now explicitly addresses this case in : §3.19.7, "ruleMessageId property":

>If the message string identified by `ruleId` and `ruleMessageId` includes placeholders (§3.9.4), then `result.message.arguments` (§3.19.6, §3.9.11) **SHALL** contain the replacement values for the placeholders. In this situation, `result.message` will contain _only_ the `arguments` property.
>
> EXAMPLE 2: In this example, the message string identified by `ruleId` and `ruleMessageId` has a single placeholder `"{0}"`. `message.arguments holds` the replacement value `"counter"`.
```
{                                # A run object (§3.11).
  "results": [
    {                            # A result object (§3.19).
      "ruleId": "CA2101",
      "ruleMessageId": "default",
      "message": {
        "arguments": [
          "counter"
        ]
      }
    }
  ],

  "resources": {
    "rules": {
      "CA2101": {                # A rule object
        "messageStrings": {
          "default": "Variable \"{0}\" is uninitialized."
        }
      }
    }
  }
}
```

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 2]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                          ],
+                          ""fixes"": [
+                            {
+                              ""description"": ""Wrap attribute values in single quotes."",
+                              ""fileChanges"": [
+                                {
+                                  ""uri"": ""http://localhost:34420/HtmlFixes.html"",
+                                  ""replacements"": [
+                                    {
+                                      ""offset"": 623,
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: For consistency: This should be 720., and the second offset should be 725.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 3]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                          ""formattedRuleMessage"": {
+                            ""formatId"": ""default"",
+                            ""arguments"": [
+                              ""shape""
+                            ]
+                          },
+                          ""suppressionStates"": [ ""suppressedExternally"" ],
+                          ""baselineState"": ""existing"",
+                          ""level"": ""error"",
+                          ""snippet"": ""<area alt=\""Here is some text\"" coords=\""10 20 20\"" href=\""moon.html\"" shape=circle xweb:fixindex=\""0\"" />"",
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: For the sake of brevity, I suggest that the example log files include only those properties that are either (a) required, (b) necessary to demonstrate the feature, or (c) helpful for understanding. In this example, `suppressionState`, `baselineState`, and `level` are optional, so I'd omit them, `snippet` helps clarify the example, so I'd keep it even though it's optional.

Other candidates for omission:
- `tool.semanticVersion` (used to be required, but is now optional)
- `run.files`

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 4]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                            {
+                              ""analysisTarget"": {
+                                ""uri"": ""http://localhost:34420/HtmlFixes.html""
+                              },
+                              ""resultFile"": {
+                                ""uri"": ""http://localhost:34420/HtmlFixes.html"",
+                                ""region"": {
+                                  ""startLine"": 24,
+                                  ""startColumn"": 4,
+                                  ""endColumn"": 38,
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: For consistency: `"endColumn": 7, "length": 3.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 5]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                          ],
+                          ""fixes"": [
+                            {
+                              ""description"": ""Convert tag name to lowercase."",
+                              ""fileChanges"": [
+                                {
+                                  ""uri"": ""http://localhost:34420/HtmlFixes.html"",
+                                  ""replacements"": [
+                                    {
+                                      ""offset"": 804,
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: For consistency: `"offset": 803`.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 1 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Great thank you.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 6]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
 
             v2LogText.Should().Be(v2LogExpectedText);
         }
+
+        [Fact]
+        public void SarifTransformerTests_ToCurrent_TwoResultsWithFixes()
+        {
+            string v1LogText =
+              @"{
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Nit: In the v2Log below, you push the sample log against the left margin. I like that better. (Saves file size, marginally easier to read, definitely easier to type).

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 2 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Fixed

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 7]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                                }
+                              ]
+                            }
+                          ]
+		                }
+                      ],
+                      ""rules"": {
+                        ""WEB1079.AttributeValueIsNotQuoted"": {
+                          ""id"": ""WEB1079"",
+                          ""shortDescription"": ""The attribute value is not quoted."",
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Another example of an optional property that aids understanding of the sample -- good.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 3 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Done

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 4 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: This is actually real data. It highlights the entire <DIV ...></DIV>

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: I can still change it of course

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 6 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Let's just fix this one for now. I'd rather not lengthen the PR. Let's do it opportunistically.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 4 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Oh! Sorry. Leave it.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 5 above)
Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Yes, now that you've explained that the region includes the whole element, this makes sense.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 6 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: OK

**Comment by EasyRhin0** [comment]
I just realized I didn't add the call to CreateFix. Kind of important for this change. :p

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 8]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                  ""length"": 5,
+                  ""snippet"": {
+                    ""text"": ""<area alt=\""Here is some text\"" coords=\""10 20 20\"" href=\""moon.html\"" shape=circle xweb:fixindex=\""0\"" />""
+                  }
+                }
+              }
+            }
+          ],
+          ""suppressionStates"": [""suppressedExternally""],
+          ""baselineState"": ""existing"",
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Bug: The fixes are not emitted at all!

**Comment by lgolding** [comment]
I will wait for your next push to review the logic.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 8 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Fixed

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 9]
```diff
@@ -1114,5 +1119,282 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+        {
+          ""ruleId"": ""WEB1079"",
+          ""level"": ""error"",
+          ""message"": {
+            ""arguments"": [
+              ""shape""
+            ]
+          },
+          ""ruleMessageId"": ""default"",
+          ""analysisTarget"": {
```
[the remark this answers is not in the document]
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Fixed

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 7 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Will keep this in mind going forward.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 10]
```diff
@@ -1114,5 +1119,325 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+          },
+          {
+            ""ruleId"": ""WEB1066"",
+            ""formattedRuleMessage"": {
+              ""formatId"": ""default"",
+              ""arguments"": [
+                ""DIV""
+              ]
+            },
+            ""suppressionStates"": [ ""suppressedExternally"" ],
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Remove these three optional attributes.

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs [hunk 11]
```diff
@@ -1114,5 +1119,325 @@ public void SarifTransformerTests_ToCurrent_BasicResult()
...
+                ""text"": ""Wrap attribute value in double quotes.""
+              },
+              ""fileChanges"": [
+                {
+                  ""fileLocation"": {
+                    ""uri"": ""http://localhost:34420/HtmlFixes.html""
+                  },
+                  ""replacements"": [
+                    {
+                      ""offset"": 720,
```
Comment (lgolding) [inline src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: In v2, a `replacement` object looks like this (§3.40, replacement object):
```
"fileChanges": [
  {
    "deletedRegion": {
      "offset": 12,
      "length": 5
    },
    "insertedContent": {
      "binary": "ZXhhbXBsZQ=="
    }
  },
  {
    "deletedRegion": {
      "offset": 20,
      "length": 3
    }
  },
  {
    "deletedRegion": {
      "offset": 312,
      "length": 0
    },
    "insertedContent": {
      "binary": "ZXhhbXBsZQ=="
    }
  }
]
```
Are we not up to date on that schema change?

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 10 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Fixed, sorry missed the second set

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 11 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Must not be.

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Confirmed. Want me to include those changes here?

Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Or I can do them separately and merge.

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Let's do this, in sequence:
1) I'll review this the way it is, and we'll merge it.
2) Separate PR: Update the schema and modify the fix transformation logic.

Comment (lgolding) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Oh, and
    3. Look for other missing schema changes. I could've sworn we were up to date.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 12]
```diff
@@ -519,21 +565,34 @@ internal Result CreateResult(ResultVersionOne v1Result)
             {
                 result = new Result
                 {
-                    AnalysisTarget = CreateFileLocation(v1Result.Locations?[0]?.AnalysisTarget),
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Bug: The fix here is not to remove `AnalysisTarget`. It's to populate it _only if it differs from the result file_.

Comment (lgolding) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Oh! I see you did that later. ok.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 13]
```diff
@@ -519,21 +565,34 @@ internal Result CreateResult(ResultVersionOne v1Result)
...
-                    RuleMessageId = v1Result.FormattedRuleMessage.FormatId,
                     Stacks = v1Result.Stacks?.Select(CreateStack).ToList(),
                     SuppressionStates = Utilities.CreateSuppressionStates(v1Result.SuppressionStates),
                 };
 
-                // TODO: transfer formatted rule messages
+                if (v1Result.Locations?[0]?.AnalysisTarget?.Uri != v1Result.Locations?[0]?.ResultFile?.Uri)
+                {
+                    result.AnalysisTarget = CreateFileLocation(v1Result.Locations[0].AnalysisTarget);
+                }
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: This is worth a comment: "The spec says that analysisTarget is required only if it differs from the result file."

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs [hunk 14]
```diff
@@ -415,6 +444,23 @@ internal Notification CreateNotification(NotificationVersionOne v1Notification)
...
+        {
+            Replacement replacement = null;
+
+            if (v1Replacement != null)
+            {
+                replacement = new Replacement
+                {
+                    DeletedLength = v1Replacement.DeletedLength,
+                    InsertedBytes = v1Replacement.InsertedBytes,
+                    Offset = v1Replacement.Offset
```
Comment (lgolding) [inline src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: In the next PR, this is the only part of the converter logic that will need to change.

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 12 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: Right, it's handled conditionally down after the result creation

File: src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs (hunk 13 above)
Comment (EasyRhin0) [reply src/Sarif/Visitors/SarifVersionOneToCurrentVisitor.cs]: OK

File: src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs (hunk 11 above)
Comment (EasyRhin0) [reply src/Sarif.UnitTests/SarifVersionOneToCurrentTests.cs]: Sounds good. I'll take 5 minutes to look back and see if it was ever changed,

**Comment by lgolding** [comment]
(oh, right, and removing that second set of optional properties)

**Comment by EasyRhin0** [comment]
Yup got those too. Off to lunch now...

**Pull request merged**